transaction-view/src/sanitize.rs

```diff
@@ -0,0 +1,393 @@
+                data: &[][..],
+                meta,
+            };
+            assert_eq!(
+                sanitize_signatures(&view),
+                Err(TransactionViewError::SanitizeError)
+            );
+        }
+
+        // Too many signatures.
+        {
+            let mut meta = dummy_metadata();
+            meta.signature.num_signatures = 2;
+            let view = UnsanitizedTransactionView {
+                data: &[][..],
+                meta,
+            };
+            assert_eq!(
+                sanitize_signatures(&view),
+                Err(TransactionViewError::SanitizeError)
+            );
+        }
+
+        // Not enough static accounts.
+        {
+            let mut meta = dummy_metadata();
+            meta.signature.num_signatures = 2;
+            meta.message_header.num_required_signatures = 2;
+            meta.static_account_keys.num_static_accounts = 1;
+            let view = UnsanitizedTransactionView {
+                data: &[][..],
+                meta,
+            };
+            assert_eq!(
+                sanitize_signatures(&view),
+                Err(TransactionViewError::SanitizeError)
+            );
+        }
+    }
+
+    #[test]
+    fn test_sanitize_account_access() {
+        // Overlap of signing and readonly non-signing accounts.
+        {
+            let mut meta = dummy_metadata();
+            meta.message_header.num_readonly_unsigned_accounts = 2;
+            meta.static_account_keys.num_static_accounts = 2;
+            let view = UnsanitizedTransactionView {
+                data: &[][..],
+                meta,
+            };
+            assert_eq!(
+                sanitize_account_access(&view),
+                Err(TransactionViewError::SanitizeError)
+            );
+        }
+
+        // Not enough writable accounts.
+        {
+            let mut meta = dummy_metadata();
+            meta.message_header.num_readonly_signed_accounts = 1;
+            meta.static_account_keys.num_static_accounts = 2;
+            let view = UnsanitizedTransactionView {
+                data: &[][..],
+                meta,
+            };
+            assert_eq!(
+                sanitize_account_access(&view),
+                Err(TransactionViewError::SanitizeError)
+            );
+        }
+
+        // Too many accounts.
+        {
+            let mut meta = dummy_metadata();
+            meta.static_account_keys.num_static_accounts = 1;
+            meta.address_table_lookup.total_writable_lookup_accounts = 200;
+            meta.address_table_lookup.total_readonly_lookup_accounts = 200;
+            let view = UnsanitizedTransactionView {
+                data: &[][..],
+                meta,
+            };
+            assert_eq!(
+                sanitize_account_access(&view),
+                Err(TransactionViewError::SanitizeError)
+            );
+        }
+    }
+
+    #[test]
+    fn test_sanitize_instructions() {
```

Comment:
It would be nice to have a test that ensures that if even when there are non static keys from ATLs, the program index still cannot be greater or equal to the number of static keys. The tests you have now don't use ATLs at all.

Reply:
Added: https://github.com/anza-xyz/agave/pull/2757/files#diff-a88f2adf2b6178065e29a7feb3fb1083b0a2282ea647ae4db1782acddb2787abR455

Reply:
I still don't see this particular test. Basically a test like your signature sanitization test "Not enough static accounts - with look up accounts" (thanks for adding that btw!) but for program id indexes

Reply:
oh it just has a slightly different comment, and the ATLs were defined above it since I clone them.

Look at the test with this comment: `// Invalid account index with v0.`

Reply:
That test doesn't have anything to do with program indexes. It just checks that account indexes can't be greater than or equal to the number of static + ATL accounts. I would like a test that shows that a program id index cannot reference an ATL account and therefore cannot be greater than or equal to the number of static accounts.

Reply:
Oh! I missed a word when reading. yeah, I get what you're saying now.

a1600ca765